include/deal.II/base/memory_space_data.h

```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
 namespace MemorySpace
 {
   /**
-   * Data structure
+   * Structure which stores data on the host or the device depending on the
+   * template parameter @p MemorySpace. The data is copied into the structure
+   * which then owns the data and will release the memory when the destructor is
+   * called.
    */
-  template <typename Number, typename MemorySpace>
+  template <typename T, typename MemorySpace>
   struct MemorySpaceData
   {
-    MemorySpaceData()
-    {
-      static_assert(std::is_same<MemorySpace, Host>::value ||
-                      std::is_same<MemorySpace, CUDA>::value,
-                    "MemorySpace should be Host or CUDA");
-    }
+    MemorySpaceData();
 
     /**
-     * Copy the active data (values for Host and values_dev for CUDA) to @p begin.
-     * If the data is on the device it is moved to the host.
+     * Copy the class member values to @p begin. If the data is on the device it is moved to the host.
      */
     void
-    copy_to(Number *begin, std::size_t n_elements)
-    {
-      (void)begin;
-      (void)n_elements;
-    }
+    copy_to(T *begin, const std::size_t n_elements);
 
     /**
-     * Copy the data in @p begin to the active data of the structure (values for
-     * Host and values_dev for CUDA). The pointer @p begin must be on the host.
+     * Copy the data in @p begin to the class member values. The pointer @p begin must be on the host.
      */
     void
-    copy_from(Number *begin, std::size_t n_elements)
-    {
-      (void)begin;
-      (void)n_elements;
-    }
+    copy_from(const T *begin, const std::size_t n_elements);
+
+    /**
+     * Kokkos View to a host buffer used for MPI communication
+     */
+    // FIXME Should we move this somewhere else?
+    Kokkos::View<T *, Kokkos::HostSpace> values_host_buffer;
 
     /**
-     * Pointer to data on the host.
+     * Kokkos View to the data on the device
      */
-    std::unique_ptr<Number[], std::function<void(Number *)>> values;
+    Kokkos::View<T *, typename MemorySpace::kokkos_space> values;
```

Comment:
Is this a view or does it also contain the actual values? I am confused by the name and how it used.

Reply:
Kokkos::View is owning unless you use an unmanaged View.  A Kokkos::View is a really some kind of shared pointer + metadata

Reply:
The number of `*`s in the first template argument gives you the dimensionality of the `View`. So far we are only using one-dimensional Views. The second template argument is the memory space. You could specify more template arguments nut they are not really needed in general.

Reply:
Sure.